frame/treasury/src/lib.rs

```diff
@@ -108,71 +145,68 @@ type AccountIdLookupOf<T> = <<T as frame_system::Config>::Lookup as StaticLookup
 ///   mark this value as `true`. This will prevent the treasury from burning the excess funds.
 #[impl_trait_for_tuples::impl_for_tuples(30)]
 pub trait SpendFunds<T: Config<I>, I: 'static = ()> {
-	fn spend_funds(
-		budget_remaining: &mut BalanceOf<T, I>,
-		imbalance: &mut PositiveImbalanceOf<T, I>,
-		total_weight: &mut Weight,
-		missed_any: &mut bool,
-	);
+	fn spend_funds(total_weight: &mut Weight, total_spent: T::Balance, total_missed: u32);
 }
 
 /// An index of a proposal. Just a `u32`.
 pub type ProposalIndex = u32;
+/// A count of proposals. Just a `u32`.
+pub type ProposalsCount = u32;
 
 /// A spending proposal.
 #[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
 #[derive(Encode, Decode, Clone, PartialEq, Eq, MaxEncodedLen, RuntimeDebug, TypeInfo)]
-pub struct Proposal<AccountId, Balance> {
+pub struct Proposal<AccountId, Balance, AssetId, SpendBalance, PaymentId> {
```

Comment:
```suggestion
pub struct Proposal<AccountId, Balance, AssetKind, AssetBalance, PaymentId> {
```

Reply:
`SpendBalance` is a bit too much like "the `Balance` which is deducted from the `spend` budget".